Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/KeyFileDownloader.kt

```diff
@@ -0,0 +1,335 @@
+package de.rki.coronawarnapp.diagnosiskeys.download
+
+import dagger.Reusable
+import de.rki.coronawarnapp.diagnosiskeys.server.DiagnosisKeyServer
+import de.rki.coronawarnapp.diagnosiskeys.server.KeyFileHeaderHook
+import de.rki.coronawarnapp.diagnosiskeys.server.LocationCode
+import de.rki.coronawarnapp.diagnosiskeys.storage.CachedKeyInfo
+import de.rki.coronawarnapp.diagnosiskeys.storage.KeyCacheRepository
+import de.rki.coronawarnapp.diagnosiskeys.storage.legacy.LegacyKeyCacheMigration
+import de.rki.coronawarnapp.storage.AppSettings
+import de.rki.coronawarnapp.storage.DeviceStorage
+import de.rki.coronawarnapp.storage.InsufficientStorageException
+import de.rki.coronawarnapp.util.CWADebug
+import de.rki.coronawarnapp.util.HashExtensions.hashToMD5
+import de.rki.coronawarnapp.util.debug.measureTimeMillisWithResult
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.withContext
+import org.joda.time.LocalTime
+import timber.log.Timber
+import java.io.File
+import javax.inject.Inject
+
+/**
+ * Downloads new or missing key files from the CDN
+ */
+@Reusable
+class KeyFileDownloader @Inject constructor(
+    private val deviceStorage: DeviceStorage,
+    private val keyServer: DiagnosisKeyServer,
+    private val keyCache: KeyCacheRepository,
+    private val legacyKeyCache: LegacyKeyCacheMigration,
+    private val settings: AppSettings
+) {
+
+    private suspend fun checkStorageSpace(countries: List<LocationCode>): DeviceStorage.CheckResult {
+        val storageResult = deviceStorage.checkSpacePrivateStorage(
+            // 512KB per day file, for 15 days, for each country ~ 65MB for 9 countries
+            requiredBytes = countries.size * 15 * 512 * 1024L
```

Comment:
magic numbers?

Reply:
Good point, extracted this and added some comments explaining the calculation.